src/types/quota.rs

```diff
@@ -0,0 +1,226 @@
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// Due to an implementation detail this is a wrapper around a single [`Quota`].
+///
+/// Used by [`Session::get_quota`] and [`Session::set_quota`]
+#[self_referencing]
+pub struct QuotaResponse {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) quota: Quota<'this>,
+}
+
+impl QuotaResponse {
+    /// Parse the [`Quota`] response from a response buffer.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        QuotaResponseTryBuilder {
+            data: owned,
+            quota_builder: |input| {
+                let mut quotas = Vec::new();
+
+                // There should only be one single QUOTA response
+                parse_many_into(input, &mut quotas, unsolicited, |response| match response {
+                    Response::Quota(q) => Ok(MapOrNot::Map(Quota::from_imap_proto(q))),
+                    resp => Ok(MapOrNot::Not(resp)),
+                })?;
+
+                match quotas.len() {
+                    1 => Ok(quotas.remove(0)),
+                    _ => Err(Error::Parse(ParseError::Invalid(input.to_vec()))),
+                }
+            },
+        }
+        .try_build()
+    }
+
+    /// Access to the wrapped [`Quota`] struct
+    pub fn quota(&self) -> &Quota<'_> {
+        self.borrow_quota()
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// Used by [`QuotaResponse`] and [`QuotaRoot`]
+#[derive(Debug, Eq, PartialEq)]
+pub struct Quota<'a> {
+    root_name: Cow<'a, str>,
+    resources: Vec<QuotaResource<'a>>,
+}
+
+impl<'a> Quota<'a> {
+    /// The quota root name
+    pub fn root_name(&self) -> &str {
+        &*self.root_name
+    }
+
+    /// The defined resources with their usage and limits (could be empty)
+    pub fn resources(&self) -> &[QuotaResource<'a>] {
+        &self.resources[..]
+    }
+}
+
+impl<'a> Quota<'a> {
+    fn from_imap_proto(q: imap_proto::Quota<'a>) -> Self {
+        Self {
+            root_name: q.root_name,
+            resources: q
+                .resources
+                .into_iter()
+                .map(|e| QuotaResource {
+                    name: match e.name {
+                        imap_proto::QuotaResourceName::Storage => QuotaResourceName::Storage,
+                        imap_proto::QuotaResourceName::Message => QuotaResourceName::Message,
+                        imap_proto::QuotaResourceName::Atom(e) => QuotaResourceName::Atom(e),
+                    },
+                    usage: e.usage,
+                    limit: e.limit,
+                })
+                .collect(),
+        }
+    }
+}
+
+/// From [QUOTA Response](https://datatracker.ietf.org/doc/html/rfc2087#section-5.1)
+///
+/// The quota resource sub-pieces in a [`Quota`]
+#[derive(Debug, Eq, PartialEq)]
+pub struct QuotaResource<'a> {
```

Comment:
Let's set this to be `#[non_exhaustive]` in case other extensions we add support for in the future might add more fields.